### ISSUE TITLE
Stream deleted but subscription stays open

### ISSUE THREAD

**Issue by scottf** [nats-io/nats-server#2682]
## Defect

Subscription is not torn down after stream is deleted.

 - [x] Included a [Minimal, Complete, and Verifiable example] (https://stackoverflow.com/help/mcve)

#### Versions of `nats-server` and affected client libraries used:

nats-server: v2.6.4

#### OS/Container environment:

Any

#### Steps or code to reproduce the issue:

1. Create a stream/subject, publish one message.
1. Open a subscription, with heartbeats, to stream/subject.
1. Receive message. Keep attempting to receive messages.
1. Delete stream with a different client.

#### Expected result:

The subscription is torn down and the client is notified, for instance through a status message, so the client can know that the subscription is terminated.

#### Actual result:

The subscription is unaffected, heartbeats continue to flow.

**Comment by ripienaar** [comment]
I cannot verify this case, I create a push consumer on the CLI, subscribe to it using a normal `nats sub` and I see heartbeats etc:

```
[#6] Received on "js.out.x.push"
Description: Idle Heartbeat
Nats-Last-Consumer: 0
Nats-Last-Stream: 0
Status: 100
```

Soon as I remove the stream the heartbeats stop. Same clustered or unclustered.

**Comment by derekcollison** [comment]
I saw inconsistent behavior that I want to look into, we attempt to clean up all consumers for a stream but did see cases where that did not seem to work 100%.

**Comment by derekcollison** [comment]
The sub on your side in the app has to be torn down there and not sure you can rely on signaling that the stream and consumer went away. The lack of HB should kick in correct behavior.

**Comment by scottf** [comment]
Maybe there could be a status message anyway, with the understanding that it may not be delivered.
Also, we don't currently require HB, so it's possible that a sub may not have. On subs w/HB we will handle it.
Also for Pull consumers, I'll assume that calling `CONSUMER.MSG.NEXT.%s.%s` will cause an api error, but I can test this. Also, if they had set a long expire or a straight batch size pull, it may be a while before they know there is a problem, so we'll need to come up with some strategies to deal with this.

**Comment by ripienaar** [comment]
Push consumers should require HB.  A status message on stream delete exist already, but attempting to rely on it for client functionality is a mistake.

**Comment by ripienaar** [comment]
Just a note - push subscribers with a qgroup probably cant have HBs, consumer health really is a big PITA.

**Comment by derekcollison** [comment]
Anyone been able to reproduce this at all? I have not been able to myself.

**Comment by scottf** [comment]
@derekcollison @aricart @matthiashanel @ripienaar
I revisited this am am closing. The heartbeats stop as they should. Don't know why I thought they continued under v2.6.4, but I just tried and verified that heartbeats stop. I also ran against v2.6.6-beta, behavior was the same.
I think we will just need to document behavior in the client, that the client consumer is alive but there are never any messages coming. This is probably an edge case anyway.
As RI mentioned, without heartbeats, there is no way to know that the stream is gone. 
So let's discuss today if we want to do something in the clients that requires heartbeats on push subscriptions and what the time for the default setting would be. It's probably a bad idea, but we could occasionally get stream or consumer info which will return `stream not found [10059]` or some other way for clients to be proactive when it's not seeing messages.
